### ISSUE TITLE
Scaling down a pod in kubernetes, causes daprd sidecar to fail all messages in queue with 500 status

### ISSUE THREAD

**Issue by WhoAteDaCake** [dapr/components-contrib#946]
## Expected Behavior
When a pod scales down, `daprd` sidecar should stop sending messages and shutdown.

## Actual Behavior
`daprd` sidecar keeps sending messages to a dead service, which causes all requests to fail with status **500**

Example with scaled 1 to 0:
![scaled](https://user-images.githubusercontent.com/16735813/122239247-b1f72300-ceb8-11eb-85a4-4e4d61a365ca.png)

## Steps to Reproduce the Problem
Scale-up any consumer to 2 pods, then scale one down all items get instantly consumed from the queue.
My configuration

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pubsub
  namespace: default
spec:
  type: pubsub.rabbitmq
  version: v1
  metadata:
  - name: host
    secretKeyRef:
      name: primary-secrets
      key: rabbitmq_url
  - name: durable
    value: "false"
  - name: deletedWhenUnused
    value: "false"
  - name: autoAck
    value: "false"
  - name: deliveryMode
    value: "0"
  - name: requeueInFailure
    value: "true"
  - name: prefetchCount
    value: "5"
  - name: reconnectWait
    value: "0"
  - name: concurrencyMode
    value: parallel
auth:
  secretStore: kubernetes
```

## Release Note

RELEASE NOTE:

**Comment by WhoAteDaCake** [comment]
Seems like quite a critical issue?

**Comment by TimVen** [comment]
If anyone is using RabbitMQ for PubSub and can scale down a pod to confirm that all messages are dropped or not that would be great!

**Comment by yaron2** [comment]
@TimVen I'll confirm that this week. Are you sure that messages are not just getting processed and thus removed from the queue?

**Comment by TimVen** [comment]
Thanks @yaron2. As soon as one of the pods is scaled down it seems to purge all queue items. We can confirm that they have not been processed.

**Comment by WhoAteDaCake** [comment]
@yaron2 I've created a repository to reproduce the issue here https://github.com/WhoAteDaCake/dapr-rabbitmq-repro-1

**Comment by Taction** [comment]
I've encountered this too, this situation only occurs when sidecar is running normally but consumer service is down or disconnected. This is because when consumers disconnect, the request from sidecar to consumer will get 500 http status code and the message will be acked in this situation.
I did two things to alleviate this problem. One is to add the [retry strategy](https://github.com/dapr/components-contrib/pull/862), and the other is to add the support for the dead letter queue(in progress). By defining a retry strategy, a short-term consumer downtime will not affect data consumption.

**Comment by WhoAteDaCake** [comment]
Thanks for the reply. Wouldn't receiving 500 only get rid of messages, which have been unacked, rather than all of the messages on the queue?

**Comment by Taction** [comment]
I am not very certain about `a consumer  unsubscribes from the queue` mean, I am just assume you are in such a situation that your consumer exited abnormally and will go back online within a period of time. During this period of time, the sidecar starts normally, and you don't want the situation you mentioned now to happen.
Adding a retry strategy can instruct the sidecar how to resend the message. During the retry of sending the message, the message will not be acked.

`backOffMaxRetries` is the maximum number of retries to process the message, set it to a bigger number will be able to tolerate longer consumer downtime. Set `backOffMaxRetries` to `"-1"` will retry indefinitely until the message is processed.(I am working on the documents about these config)
You can try the following config. And this feature is not included in dapr version 1.2.2, which means you must build it from source at master branch.
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pubsub
  namespace: default
spec:
  type: pubsub.rabbitmq
  version: v1
  metadata:
    - name: host
      secretKeyRef:
        name: primary-secrets
        key: rabbitmq_url
    - name: durable
      value: "false"
    - name: deletedWhenUnused
      value: "false"
    - name: autoAck
      value: "false"
    - name: deliveryMode
      value: "0"
    - name: requeueInFailure
      value: "true"
    - name: prefetchCount
      value: "5"
    - name: reconnectWait
      value: "0"
    - name: concurrencyMode
      value: parallel
    # backoff option
    - name: backOffPolicy
      value: "exponential"
    - name: backOffMaxRetries
      value: "10"
auth:
  secretStore: kubernetes
```

**Comment by WhoAteDaCake** [comment]
Thanks for the solution. The situation we are in is when a pod is scaled down, the primary service goes down first, which causes daprd sidecar to send a barrage of requests all of them returning status 500.  Thanks to your notes, I've looked into the logs of `daprd` container further and found:

> dapr shutting down. Waiting 5s to finish outstanding operations
